modules/parent-join/src/main/java/org/elasticsearch/join/aggregations/ChildrenToParentAggregator.java

```diff
@@ -0,0 +1,216 @@
+        final SortedSetDocValues globalOrdinals = valuesSource.globalOrdinalsValues(ctx);
+        final Bits childDocs = Lucene.asSequentialAccessBits(ctx.reader().maxDoc(), childFilter.scorerSupplier(ctx));
+        return new LeafBucketCollector() {
+
+            @Override
+            public void collect(int docId, long bucket) throws IOException {
+                if (childDocs.get(docId) && globalOrdinals.advanceExact(docId)) {
+                    long globalOrdinal = globalOrdinals.nextOrd();
+                    assert globalOrdinals.nextOrd() == SortedSetDocValues.NO_MORE_ORDS;
+                    if (globalOrdinal != -1) {
+                        if (childrenOrdToBuckets.get(globalOrdinal) == -1) {
+                            childrenOrdToBuckets.set(globalOrdinal, bucket);
+                        } else {
+                            long[] bucketOrds = childrenOrdToOtherBuckets.get(globalOrdinal);
+                            if (bucketOrds != null) {
+                                bucketOrds = Arrays.copyOf(bucketOrds, bucketOrds.length + 1);
+                                bucketOrds[bucketOrds.length - 1] = bucket;
+                                childrenOrdToOtherBuckets.put(globalOrdinal, bucketOrds);
+                            } else {
+                                childrenOrdToOtherBuckets.put(globalOrdinal, new long[] { bucket });
+                            }
+                            multipleBucketsPerChildrenOrd = true;
+                        }
+                    }
+                }
+            }
+        };
+    }
+
+    @Override
+    protected void doPostCollection() throws IOException {
+        IndexReader indexReader = context().searcher().getIndexReader();
+        for (LeafReaderContext ctx : indexReader.leaves()) {
+            Scorer parentDocsScorer = parentFilter.scorer(ctx);
+            if (parentDocsScorer == null) {
+                continue;
+            }
+            DocIdSetIterator parentDocsIter = parentDocsScorer.iterator();
+
+            final LeafBucketCollector sub = collectableSubAggregators.getLeafCollector(ctx);
+
+            final SortedSetDocValues globalOrdinals = valuesSource.globalOrdinalsValues(ctx);
+            // Set the scorer, since we now replay only the parent docIds
+            sub.setScorer(new Scorable() {
+                @Override
+                public float score() {
+                    return 1f;
+                }
+
+                @Override
+                public int docID() {
+                    return parentDocsIter.docID();
+                }
+            });
+
+            // TODO: this is unwanted allocation, just for initial verification
+            // of the implementation idea, probably needs to be done differently
+            // for production use so we do not cause memory churn here
+            //Set<Tuple<Integer,Long>> seenParents = new HashSet<>();
+            Map<Integer, Set<Long>> seenParents = new HashMap<>();
```

Comment:
I don't think this is needed because we can have only one parent per global ordinal ? The loop below iterates over the matching parent so you shouldn't see the same global ordinal more than once.

Reply:
Somehow I needed it to get correct results, but it seems I can remove it as soon as I do a check for existing buckets in the LeafCollector.